roles/oVirt.datacenter-cleanup/tasks/main.yml

```diff
@@ -11,27 +14,29 @@
 - name: Remove Disks
   include: disks.yml
 
-- name: Find existing Storage Domains in Datacenter
-  ovirt_storage_domains_facts:
-    auth: "{{ ovirt_auth }}"
-    pattern: "datacenter={{ data_center_name }}"
-
-- name: Remove all Storage Domains except master
+- name: Remove all Storage Domains except master and hosted_storage if HE
   include: storages_pre.yml
 
 - name: Find existing clusters in Datacenter
-  ovirt_clusters_facts:
+  ovirt_cluster_facts:
     auth: "{{ ovirt_auth }}"
     pattern: "datacenter={{ data_center_name }}"
 
+- name: Remove hosts except from the one hosting the engine VM
+  include: remove_hosts_not_hosting_engine_vm.yml
+  when: hosted_engine
+
 - name: Remove Datacenter
   include: datacenter.yml
+  when: not hosted_engine
 
 - name: Remove master Storage Domain
   include: storages_last.yml
+  when: not hosted_engine
 
 - name: Remove Clusters and Hosts
   include: cluster_and_hosts.yml
   with_items: "{{ ovirt_clusters }}"
   loop_control:
     loop_var: cluster_item
+  when: hosted_engine_cluster_id and cluster_item.id != hosted_engine_cluster_id
```

Comment:
I think you can do it same as in file roles/oVirt.datacenter-cleanup/tasks/remove_hosts_not_hosting_engine_vm.yml line 25.  Or you can remove default value and use when: hosted_engine_cluster_id is defined and cluster_item.id != hosted_engine_cluster_id then it could also fix the issue here.

Reply:
default('') will be evaluated as not defined ... so it is the same.

Reply:
empty string is evaluated as defined variable..  so once you initialize some var it's evaluated as defined. But once it's not defined you can prevent fail your code when first part of condition is:
when: hosted_engine_cluster_id is defined and another_condition = hosted_engine_cluster_id
This case you won't fail here.

As you can see I was failing here with:
```bash
13:38:45 TASK [oVirt.infra/roles/oVirt.datacenter-cleanup : Remove Clusters and Hosts] ***
13:38:45 fatal: [localhost]: FAILED! => {"msg": "The conditional check 'hosted_engine_cluster_id and cluster_item.id != hosted_engine_cluster_id' failed. The error was: error while evaluating conditional (hosted_engine_cluster_id and cluster_item.id != hosted_engine_cluster_id): 'hosted_engine_cluster_id' is undefined\n\nThe error appears to have been in '/var/lib/jenkins/workspace/rhv-ge-ansible-cleaner-4.2/ansible-playbooks/roles/oVirt.infra/roles/oVirt.datacenter-cleanup/tasks/main.yml': line 37, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Remove Clusters and Hosts\n  ^ here\n"}
```